[Diagnostics] Extend single parameter tuple splat to support generic parameters

xedin opened the pull request:
Currently single parameter tuple splat fix/diagnostic supports only
cases where parameter is a concrete tuple type, let's enhance that to
support generic parameters as well e.g.:

```swift
func foo<T>(_: T) {}
foo(0, 1, 2) // `T` expects arguments to form a tuple e.g. `foo((0, 1, 2))`
```

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[Diagnostics] Extend single parameter tuple splat to support generic parameters

Currently single parameter tuple splat fix/diagnostic supports only
cases where parameter is a concrete tuple type, let's enhance that to
support generic parameters as well e.g.:

```swift
func foo<T>(_: T) {}
foo(0, 1, 2) // `T` expects arguments to form a tuple e.g. `foo((0, 1, 2))`
```

xedin committed:
[Diagnostics] Fix fix-it location for labeled tuple splat

In situations like this:

```swift
func foo(x: (Int, Int) {}
foo(x: 0, 0)
```

Left paren to form a missing tuple should be placed after
the label because belongs to the parameter and not the tuple.

xedin commented:
@swift-ci please smoke test

jrose-apple commented on test/Constraints/tuple_arguments.swift:
```diff
@@ -256,12 +256,12 @@ do {
 
   s.generic(3)
-  s.generic(3, 4) // expected-error {{extra argument in call}}
+  s.generic(3, 4) // expected-error {{instance method 'generic' expects a single parameter of type 'T' [with T = (Int, Int)]}} {{13-13=(}} {{17-17=)}}
```
Eek. This definitely feels worse to me. Maybe if it were just a note, so it wasn't like we were saying this is necessarily the *right* fix.

xedin replied:
Why shouldn't we say that it's the right fix? That's how the test on line 261 deals with the problem. What are the other ways to fix this besides removing an argument?

jrose-apple replied:
Removing an argument is more likely to be the right fix. Someone is way more likely to have mistaken the signature of the function than to have been trying to pass a tuple to it.

jrose-apple replied:
The main exceptions to this were places where the generic argument was specified or inferred by something other than "the argument list is too big".

xedin replied:
I'm not sure based on the fact that all of the examples in this file fix the problem by adding parens around arguments to form a tuple. It might be synthetic, but I think it's reasonable to say that function expects a single parameter which means that you have to form a tuple to fix that. It could be that signature of the function was mistaken to be `<T>(_: T...)`. Also it might make sense to have this behavior be consistent with how labels are treated - having an argument is evidence enough that it's intended, so removing it is the last thing we'd want to do.

xedin replied:
This example seems similar - https://github.com/apple/swift/pull/27584/files#diff-d7745944dfea01aea2e14a30073bf084R107 `Left` accepts a single generic parameter and contextually it would be inferred as a tuple but arguments are distinct. I guess code like that would not be used standalone but will be similar to example in `enum_cases.swift` where it has some contextual information that suggests the shape.

jrose-apple replied:
This file is definitely synthetic. We're also further and further from the days of Swift 3. I wouldn't necessarily make a fix-it to remove the argument (because which argument would you remove?) but neither do I think we should assume that a tuple was intended.

xedin replied:
Should it just not produce a fix-it at all then in cases where single parameter is a generic parameter, but keep the same diagnostic?

xedin replied:
I have pushed an alternative change - `did you mean to pass a tuple?` from your original suggestion, but instead of making it a note I just made it a part of the diagnostic itself, just like we do for other error messages. WDYT?

jrose-apple replied:
I still don't think it's as good, especially since to a first approximation people don't read and will just choose the fix-it. I know they can do that with a note too, but still: a fix-it directly on the diagnostic is supposed to indicate that it's the only reasonable solution.

xedin replied:
Done!

xedin committed:
[Diagnostics] Clarify tuple splat message when single parameter is generic parameter

Since it's hard to say whether tuple use is really intended, let's
add `did you mean to pass a tuple?` note to the error message.

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test
